test/Attenuation_test.cpp

```diff
@@ -0,0 +1,180 @@
+
+    osg::Geode *cone = new osg::Geode();
+    cylinder->addDrawable(new osg::ShapeDrawable(new osg::Cone(osg::Vec3(0, 30, 0), 10, 10)));
+    root->addChild(cone);
+
+    osg::Geode *box = new osg::Geode();
+    cylinder->addDrawable(new osg::ShapeDrawable(new osg::Box(osg::Vec3(0, -30, -10), 10)));
+    root->addChild(box);
+}
+
+// define different point of views of the same scene
+void viewPointsFromScene(std::vector<osg::Vec3d> *eyes,
+                          std::vector<osg::Vec3d> *centers,
+                          std::vector<osg::Vec3d> *ups) {
+
+    // point of view 1 - near from the ball with the cylinder in back
+    eyes->push_back(osg::Vec3d(-8.77105, -4.20531, -3.24954));
+    centers->push_back(osg::Vec3d(-7.84659, -4.02528, -2.91345));
+    ups->push_back(osg::Vec3d(-0.123867, -0.691871, 0.711317));
+
+    // point of view 2 - near from the ball with the cube in back
+    eyes->push_back(osg::Vec3d(3.38523, 10.093, 1.12854));
+    centers->push_back(osg::Vec3d(3.22816, 9.12808, 0.918259));
+    ups->push_back(osg::Vec3d(-0.177264, -0.181915, 0.967204));
+
+    // point of view 3 - near the cone in up side
+    eyes->push_back(osg::Vec3d(-10.6743, 38.3461, 26.2601));
+    centers->push_back(osg::Vec3d(-10.3734, 38.086, 25.3426));
+    ups->push_back(osg::Vec3d(0.370619, -0.854575, 0.36379));
+
+    // point of view 4 - Faced the cube plane
+    eyes->push_back(osg::Vec3d(0.0176255, -56.5841, -10.0666));
+    centers->push_back(osg::Vec3d(0.0176255, -55.5841, -10.0666));
+    ups->push_back(osg::Vec3d(0, 0, 1));
+}
+
+BOOST_AUTO_TEST_CASE(applyShaderNormalDepthMap_TestCase) {
+    // sonar parameters
+    float maxRange = 50;            // 50 meters
+    float fovX = M_PI * 1.0 / 6;    // 30 degrees
+    float fovY = M_PI * 1.0 / 6;    // 30 degrees
+    uint height = 500;
+
+    // attenuation coefficient
+    double frequency = 700.0;   // kHz
+    double temperature = 20.0;  // celsius degrees
+    double depth = 1;           // meters
+    double salinity = 0;        // ppt
+    double acidity = 8;         // pH
+    double attenuationCoeff = underwaterSignalAttenuation(frequency, temperature, depth, salinity, acidity);
+
+     // define the different camera point of views
+    std::vector<osg::Vec3d> eyes, centers, ups;
+    viewPointsFromScene(&eyes, &centers, &ups);
+
+    // create a simple scene with multiple objects
+    osg::ref_ptr<osg::Group> root = new osg::Group();
+    makeSimpleScene(root);
+
+    // display the same with and without underwater acoustic attenuation
+    for (uint i = 0; i < eyes.size(); ++i) {
+        cv::Mat raw_shader = computeNormalDepthMap(root, height, maxRange, fovX, fovY, 0, eyes[i], centers[i], ups[i]);
+        cv::Mat att_shader = computeNormalDepthMap(root, height, maxRange, fovX, fovY, attenuationCoeff, eyes[i], centers[i], ups[i]);
+        plotSonarTest(raw_shader, maxRange, fovX * 0.5);
+        plotSonarTest(att_shader, maxRange, fovX * 0.5);
+  }
+}
```

Comment:
How do you test the code are correctly?
You just plot the final image and check visually?
IMO, you should choose some reference points in final image (sonar plot or raw shader image) and compare these points with reference value. This way you can track and sense if any code change gives the same expected result.
See the normalDepthMap_test.cpp.

Reply:
Ok. I will add this check.